### PR TITLE
added Gaussian compressor and the testing file

### PR DESCRIPTION
**Pull request by TongtongJin**
The pull request includes two new files:

(1) Compressors/Gaussian.jl': Includes implementation of structures and multiplication functions for Gaussian compression method.

(2) test/Compressors/Gaussian.jl: Includes the tests for the multiplication functions, the data structures and the update/complete functions.

### CODE REVIEW

File: src/Compressors.jl [hunk 1]
```diff
@@ -387,10 +387,10 @@ end
 
 function mul!(
     C::AbstractMatrix,
-    S::CompressorAdjoint{<:CompressorRecipe},
+    S::CompressorAdjoint,
     A::AbstractMatrix,
-    alpha,
-    beta,
+    alpha::Float64,
+    beta::Float64,
```
Comment (vp314) [inline src/Compressors.jl:393]: These have been updated to ::Number

Comment (vp314) [inline src/Compressors.jl:393]: These are updated to `Number` types in Compressor SparseSign

File: src/Compressors.jl [hunk 2]
```diff
@@ -401,9 +401,9 @@ end
 function mul!(
     C::AbstractMatrix,
     A::AbstractMatrix,
-    S::CompressorAdjoint{<:CompressorRecipe},
-    alpha,
-    beta,
+    S::CompressorAdjoint,
+    alpha::Float64,
+    beta::Float64,
```
Comment (vp314) [inline src/Compressors.jl:406]: These should be updated to `Number`

File: Project.toml [hunk 3]
```diff
@@ -1,13 +1,6 @@
...
 uuid = "da1d211a-c3bc-43b2-b5d0-ce83f16a0f3a"
-authors = [
-    "Christian Varner <cvarner@wisc.edu>", 
-    "Daniel Adrian Maldonado <maldonadod@anl.gov>", 
-    "Nathaniel Pritchard <Nathaniel.Pritchard@maths.ox.ac.uk>", 
-    "TongTong Jin <ttjin@uchicago.edu>", 
-    "Tunan Wang <tunan.wang@wisc.edu>", 
-    "Vivak Patel <vivak.patel@wisc.edu>"
-]
+authors = ["Christian Varner <cvarner@wisc.edu>", "Daniel Adrian Maldonado <maldonadod@anl.gov>", "Nathaniel Pritchard <Nathaniel.Pritchard@maths.ox.ac.uk>", "TongTong Jin <ttjin@uchicago.edu>", "Tunan Wang <tunan.wang@wisc.edu>", "Vivak Patel <vivak.patel@wisc.edu>"]
```
Comment (vp314) [inline Project.toml:3]: This is the wrong format.

File: src/Compressors/Gaussian.jl [hunk 4]
```diff
@@ -0,0 +1,98 @@
+struct Gaussian <: Compressor
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:1]: Needs a docstring

File: src/Compressors/Gaussian.jl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    n_rows::Int64
+    n_cols::Int64
+end
+
+function Gaussian(;n_rows::Int64 = 0, n_cols::Int64 = 0)
+    # Partially construct the Gaussian datatype
+    return Gaussian(n_rows, n_cols)
+end
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:9]: Constructors require a docstring. Defaults should not be zero.

File: src/Compressors/Gaussian.jl [hunk 6]
```diff
@@ -0,0 +1,98 @@
+struct Gaussian <: Compressor
+    n_rows::Int64
+    n_cols::Int64
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:3]: Should have an internal constructor that checks the validity of the input arguments.

File: src/Compressors/Gaussian.jl [hunk 7]
```diff
@@ -0,0 +1,98 @@
+struct Gaussian <: Compressor
+    n_rows::Int64
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:2]: Should target data type @nathanielpritchard be a field?

File: src/Compressors/Gaussian.jl [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+function Gaussian(;n_rows::Int64 = 0, n_cols::Int64 = 0)
+    # Partially construct the Gaussian datatype
+    return Gaussian(n_rows, n_cols)
+end
+
+mutable struct GaussianRecipe <: CompressorRecipe
+    n_rows::Int64
+    n_cols::Int64
+    scale::Float64
+    sketch_matrix::Matrix{Float64}
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:15]: I think we should let GaussianRecipe be a parametric type with the target data type specified by the user.

File: src/Compressors/Gaussian.jl [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            sample_size = n_cols
+        elseif n_cols == size(A, 2)
+            initial_size = n_cols
+            sample_size = n_rows
+        else
+            @assert false "Either you inputted row or column dimension must match \\
+            the column or row dimension of the matrix."
+        end
+
+    end
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:51]: Dispatching off of cardinality as in SparseSign is probably more appropriate here.

File: src/Compressors/Gaussian.jl [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        if n_rows == size(A, 2)
+            initial_size = n_rows
+            sample_size = n_cols
+        elseif n_cols == size(A, 2)
+            initial_size = n_cols
+            sample_size = n_rows
+        else
+            @assert false "Either you inputted row or column dimension must match \\
+            the column or row dimension of the matrix."
+        end
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:49]: Should throw an error. We should avoid `@assert`.

File: src/Compressors/Gaussian.jl [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    # Partially construct the Gaussian datatype
+    return Gaussian(n_rows, n_cols)
+end
+
+mutable struct GaussianRecipe <: CompressorRecipe
+    n_rows::Int64
+    n_cols::Int64
+    scale::Float64
+    sketch_matrix::Matrix{Float64}
+end
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:16]: Terminology should be kept the same across compression methods. E.g., `op` in SparseSign, whereas here it is `sketch_matrix`

File: src/Compressors/Gaussian.jl (hunk 8 above)
Comment (TongtongJin) [reply src/Compressors/Gaussian.jl:15]: Do you mean that we define the type of scale and eltype of sketch matrix the same as the eltype of A (let's say T) and then generate the Gaussian sketch matrix by scale .* randn(T, n_rows, n_cols)?

File: src/Compressors/Gaussian.jl [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+- `cardinality::Type{<:Cardinality}`, the direction the compression matrix is intended to be 
+    applied to a target matrix or operator. Values allowed are `Left` or `Right`. By default 
+    `Left` is chosen. 
+- `compression_dim`, the target compression dimension. Referred to as ``s`` in the
+    mathemtical description. By default this is set to 2.
+
+## Returns 
+- A `Gaussian` object. 
+"""
+
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:45]: No white space

File: src/Compressors.jl [hunk 13]
```diff
@@ -430,6 +433,48 @@ function mul!(
     return nothing
 end
 
+function mul!(x::AbstractVector, S::CompressorAdjoint, y::AbstractVector)
+    mul!(x, S, y, 1.0, 0.0)
```
Comment (Tuna2222) [inline src/Compressors.jl:437]: Why the code coverage gives warning? See line 46 in `multiplication_error_tests.jl`

File: test/Compressors/Gaussian.jl [hunk 14]
```diff
@@ -0,0 +1,157 @@
+module Gaussian_compressor
+    using Test, RLinearAlgebra
+    import Base.:*
+    import LinearAlgebra: transpose, adjoint
```
Comment (Tuna2222) [inline test/Compressors/Gaussian.jl:4]: Minor things: maybe the imports of transpose and adjoint are not used here

File: test/Compressors/Gaussian.jl [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    import LinearAlgebra: mul!, lmul!
+    import Random: randn!, seed!
+    using ..FieldTest
+    using ..ApproxTol
```
Comment (Tuna2222) [inline test/Compressors/Gaussian.jl:8]: When I use `include("test/compressors/Gaussian.jl")`, there throws the error: `LoadError: UndefVarError: `FieldTest` not defined in `Main``.

File: src/Compressors.jl [hunk 16]
```diff
@@ -15,6 +15,28 @@ compression technique to a particular set of matrices and vectors.
...
+An abstract type specifying that a matrix should be applied from the left.
+"""
+abstract type Left <: Cardinality end
+
+"""
+    Right <: Cardinality
+
+An abstract type specifying that a matrix should be applied from the right.
+"""
+abstract type Right <: Cardinality end
```
Comment (vp314) [inline src/Compressors.jl:38]: These have been changed to `struct`s in the sparse sign branch

File: src/Compressors.jl [hunk 17]
```diff
@@ -147,8 +167,10 @@ $(comp_method_description[:update_compressor])
 - Returns `nothing` 
 """
 function update_compressor!(S::CompressorRecipe)
-    throw(ArgumentError("No method `update_compressor` exists for compressor recipe of type\
-    $(typeof(S))."))
+    throw(
+        ArgumentError("No method `update_compressor` exists for compressor recipe of type\
+  $(typeof(S)).")
```
Comment (vp314) [inline src/Compressors.jl:172]: Very strange indenting...

File: src/Compressors/Gaussian.jl [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+## Arguments 
+- `cardinality::Type{<:Cardinality}`, the direction the compression matrix is intended to be 
+    applied to a target matrix or operator. Values allowed are `Left` or `Right`. By default 
+    `Left` is chosen. 
+- `compression_dim`, the target compression dimension. Referred to as ``s`` in the
+    mathemtical description. By default this is set to 2.
+
+## Returns 
+- A `Gaussian` object.
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:43]: needs to be updated with Sparse Sign's usage of structs for Left and Right

File: src/Compressors/Gaussian.jl [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+- `compression_dim`, the target compression dimension. Referred to as ``s`` in the
+    mathemtical description. By default this is set to 2.
+
+## Returns 
+- A `Gaussian` object. 
+"""
+struct Gaussian <: Compressor
+    cardinality::Type{<:Cardinality}
+    compression_dim::Int64
+    # Check on the compression dimension
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:48]: needs `type::Type{<:Number}` or `type::Type{N} where N<:Number`

File: src/Compressors/Gaussian.jl [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+    function Gaussian(cardinality, compression_dim)
+        if compression_dim < 0
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:50]: This only checks if `compression_dim` is negative, not if it is positive. Should be
`compression_dim < 1`

File: test/Compressors/Gaussian.jl [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+
+    @testset "Compressor_Gaussian" begin
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:10]: Test name needs to be updated to match Sparse Sign style

File: test/Compressors/Gaussian.jl [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        @test_compressor GaussianRecipe
+        let 
+            seed!(21321)
+            S1 = Gaussian()
+            @test typeof(S1) <: Compressor
+            @test S1.cardinality == Left
+            @test S1.compression_dim == 2
+            # Test the case when the compression_dim input is not positive
+            @test_throws ArgumentError Gaussian(cardinality = Left, compression_dim = -7)
+        end
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:20]: May want to separate these out into two blocks as they are testing for different things. The first one is testing for default behavior of the constructor while the second is constructing a new Compressor to see if it throws an error. Should be testing the edge case (i.e., `compression_dim=0`) as well

File: test/Compressors/Gaussian.jl [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+        
+        # Test the left compressor recipe construction and updating wirh default parameters
+        let
+            seed!(21321)
+            S1 = Gaussian()
+            @test typeof(S1) <: Compressor
+            @test S1.cardinality == Left
+            @test S1.compression_dim == 2
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:28]: This is already tested

File: test/Compressors/Gaussian.jl [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+            # Test completion of constructor
+            A = rand(4, 2)
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:30]: the dims of A can be passed as parameters to the let block, makes modifying tests easier. Even A can be passed as parameter to the let block after the dims of A are defined

File: src/Compressors/Gaussian.jl [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+    cardinality::Type{<:Cardinality}
+    compression_dim::Int64
+    n_rows::Int64
+    n_cols::Int64
+    scale::Float64
+    op::Matrix{Float64}
+end
+
+function complete_compressor(
+    ingredients::Gaussian, A::AbstractMatrix, type::DataType=eltype(A)
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:87]: `type` here should not be an argument if `type` is specified as a field of `ingredients`

File: src/Compressors/Gaussian.jl [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+        n_cols = size(A, 1)
+        initial_size = n_cols
+    else
+        n_rows = size(A, 2)
+        n_cols = ingredients.compression_dim
+        initial_size = n_rows
+    end
+
+    # Generate entry values by N(0,1/d)
+    T = type
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:100]: Why note just call `type` instead of `T`?

File: test/Compressors/Gaussian.jl [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+            @test S_method.n_cols == size(A,1)
+            sample_size = S_method.compression_dim
+            @test S_method.scale == 1 / sqrt(sample_size)
+            @test typeof(S_method.op) <: Matrix{eltype(A)}
+            @test size(S_method.op, 2) == size(A, 1)
+            # Test that update_compressor updates the entries
+            sketch_matrix = deepcopy(S_method.op)
+            update_compressor!(S_method)
+            # Check that at least one of the entries is different
+            @test sum(sketch_matrix .== S_method.op) < 4 * 2
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:46]: This should be its own block as it is testing something distinct from construction.

I am unclear on the goal of line 46.

File: test/Compressors/Gaussian.jl [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+            sample_size = S_method.compression_dim
+            @test S_method.scale == 1 / sqrt(sample_size)
+            @test typeof(S_method.op) <: Matrix{eltype(A)}
+            @test size(S_method.op, 1) == size(A, 2)
+            # Test that update_compressor updates the entries
+            sketch_matrix = deepcopy(S_method.op)
+            update_compressor!(S_method)
+            # Check that at least one of the entries is different
+            @test sum(sketch_matrix .== S_method.op) < 6 * 2 
+        end
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:74]: Same comments from `Left` case apply here

File: test/Compressors/Gaussian.jl [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+            S_method = complete_compressor(S1, A)
+            @test typeof(S_method) <: CompressorRecipe
+            # Check that default values are correct and appropriate allocations have been made
+            @test S_method.cardinality == Left
+            @test S_method.compression_dim == 2
+            @test S_method.n_rows == S1.compression_dim
+            @test S_method.n_cols == size(A,1)
+            sample_size = S_method.compression_dim
+            @test S_method.scale == 1 / sqrt(sample_size)
+            @test typeof(S_method.op) <: Matrix{eltype(A)}
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:40]: Maybe be explicit that is should be `==` rather than testing for a subtype

File: test/Compressors/Gaussian.jl [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+            S_method = complete_compressor(S2, A)
+            @test typeof(S_method) <: CompressorRecipe
+            # Check that default values are correct and appropriate allocations have been made
+            @test S_method.cardinality == Right
+            @test S_method.compression_dim == 2
+            @test S_method.n_rows == size(A,2)
+            @test S_method.n_cols == S2.compression_dim
+            sample_size = S_method.compression_dim
+            @test S_method.scale == 1 / sqrt(sample_size)
+            @test typeof(S_method.op) <: Matrix{eltype(A)}
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:67]: You should change this to `==` to check that it is for a type

File: test/Compressors/Gaussian.jl [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+            sketch_size = 6
+            A = rand(n_rows, n_cols)
+            B = rand(sketch_size, n_cols)
+            C1 = rand(sketch_size, n_cols)
+            C2 = rand(n_rows, sketch_size)
+            C3 = rand(n_rows, n_cols)
+            x = rand(n_rows)
+            y = rand(sketch_size)
+            S_info = Gaussian(cardinality = Left, compression_dim = sketch_size)
+            S = complete_compressor(S_info, A)
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:90]: Assuming successful tests, most of these should be parameters into the let block

File: test/Compressors/Gaussian.jl [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+            
+            # Form a vector corresponding to the columns to generate the sketch matrix
+            S_test = S.op
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:93]: This should be deepcopy? Presumably, S.op is not being changed by any of the multplications.

File: test/Compressors/Gaussian.jl [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+            mul!(C3', B', S, 2.0, 0.0)
+            @test C3' ≈ 2.0 * B' * S_test 
+            mul!(y, S, x, 2.0, 0.0)
+            @test y ≈ 2.0 * S_test * x 
+            mul!(x, S', y, 2.0, 0.0)
+            @test x ≈ 2.0 * S_test' * y 
+        end
+
+        # Test the different multiplications with the right
+        let
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:115]: Same comments apply here from `Left` case

**Review by vp314 — CHANGES_REQUESTED** [review]
Overall, a nice implementation. There are a few details that need to be changed to align with SparseSign, and a few stylistic changes requested.

File: test/Compressors/Gaussian.jl (hunk 21 above)
Comment (TongtongJin) [reply test/Compressors/Gaussian.jl:10]: changed the name into "Gaussia"

File: src/Compressors/Gaussian.jl (hunk 19 above)
Comment (TongtongJin) [reply src/Compressors/Gaussian.jl:48]: Added type as a component

File: test/Compressors/Gaussian.jl (hunk 14 above)
Comment (TongtongJin) [reply test/Compressors/Gaussian.jl:4]: Deleted import transpose, adjoint

File: src/Compressors/Gaussian.jl [hunk 34]
```diff
@@ -0,0 +1,169 @@
...
+# Constructor
+
+    Gaussian(;cardinality=Left(), compression_dim=2, type=Float64)
+
+## Arguments
+
+  - `carinality::Cardinality`, the direction the compression matrix is intended to be
+    applied to a target matrix or operator. Values allowed are `Left()` or `Right()`.
+    By default `Left()` is chosen.
+  - `compression_dim`, the target compression dimension. Referred to as ``s`` in the
```
Comment (Tuna2222) [inline src/Compressors/Gaussian.jl:42]: Maybe `compression_dim::Int64`. A naive question: "why we need both `Fields` and `Arguments`?"

File: src/Compressors/Gaussian.jl [hunk 35]
```diff
@@ -0,0 +1,169 @@
...
+    y::AbstractVector,
+    alpha::Number,
+    beta::Number,
+)
+    # Check the compatability of the sizes of the things being multiplied
+    left_mul_dimcheck(x, S, y)
+    mul!(x, S.parent.op', y, alpha, beta)
+    return nothing
+end
+
```
Comment (dmaldona) [inline src/Compressors/Gaussian.jl:149]: @TongtongJin we do not need vector mul versions

**Review by dmaldona — CHANGES_REQUESTED** [review]

File: src/Compressors/Gaussian.jl (hunk 35 above)
Comment (nathanielpritchard) [reply src/Compressors/Gaussian.jl]: done

File: Project.toml [hunk 36]
```diff
@@ -11,7 +11,9 @@ authors = [
 version = "0.2.1"
 
 [deps]
+JuliaFormatter = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
```
[the remark this answers is not in the document]
Comment (nathanielpritchard) [reply Project.toml]: done

File: src/Compressors/Gaussian.jl (hunk 35 above)
Comment (dmaldona) [reply src/Compressors/Gaussian.jl]: Great

File: src/Compressors/Gaussian.jl [hunk 37]
```diff
@@ -0,0 +1,154 @@
+"""
+    Gaussian <: Compressor
+
+An implementation of the Gaussian compression method. This method forms a Gaussian sketch matrix
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:4]: This line is beyond the allowed character limit

File: src/Compressors/Gaussian.jl [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+The recipe containing all allocations and information for the Gaussian compressor.
+
+# Fields
+
+  - `cardinality::C where C<:Cardinality`, the cardinality of the compressor. The
+    value is either `Left()` or `Right()`.
+  - `compression_dim::Int64`, the target compression dimension.
+  - `n_rows::Int64`, the number of rows of the compression matrix.
+  - `n_cols::Int64`, the number of columns of the compression matrix.
+  - `scale::Number`, the standard deviation of Gaussian distribution during the compression matrix generation.
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:87]: This line is beyond the character limit

File: src/Compressors/Gaussian.jl [hunk 39]
```diff
@@ -0,0 +1,154 @@
+"""
+    Gaussian <: Compressor
+
+An implementation of the Gaussian compression method. This method forms a Gaussian sketch matrix
+with number of rows or number of columns setting as the compression dimension.
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:5]: A simple description may be better. E.g., `A specification for a Gaussian compressor.`

File: src/Compressors/Gaussian.jl [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+        if compression_dim <= 0
+            throw(ArgumentError("Field 'Compression_dim' must be positive."))
+        end
+
+        return new(cardinality, compression_dim, type)
+    end
+end
+
+function Gaussian(;
+    cardinality::Cardinality=Left(), compression_dim::Int64=2, type::Type{<:Number}=Float64
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:69]: The arguments should be on individual lines to stay with the `Blue` style.

File: src/Compressors/Gaussian.jl [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+  - `ArgumentError` if `compression_dim` is non-positive
+"""
+struct Gaussian <: Compressor
+    cardinality::Cardinality
+    compression_dim::Int64
+    type::Type{<:Number}
+    # Check on the compression dimension
+    function Gaussian(cardinality, compression_dim, type)
+        if compression_dim <= 0
+            throw(ArgumentError("Field 'Compression_dim' must be positive."))
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:61]: `Compresion_dim` -> `compression_dim`

File: src/Compressors/Gaussian.jl [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+function GaussianRecipe(cardinality::Right, compression_dim::Int64, type::Type{<:Number}, A::AbstractMatrix)
+    n_rows = size(A, 2)
+    n_cols = compression_dim
+    initial_size = n_rows
+    # Generate entry values by N(0,1/d)
+    scale = convert(type, 1 / sqrt(compression_dim))
+    op = scale .* randn(type, n_rows, n_cols)
+    return GaussianRecipe(
+        cardinality, compression_dim, n_rows, n_cols, scale, op
+    )
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:108]: Should be on one line

File: src/Compressors/Gaussian.jl [hunk 43]
```diff
@@ -0,0 +1,154 @@
...
+function GaussianRecipe(cardinality::Left, compression_dim::Int64, type::Type{<:Number}, A::AbstractMatrix)
+    n_rows = compression_dim
+    n_cols = size(A, 1)
+    initial_size = n_cols
+    # Generate entry values by N(0,1/d)
+    scale = convert(type, 1 / sqrt(compression_dim))
+    op = scale .* randn(type, n_rows, n_cols)
+    return GaussianRecipe(
+        cardinality, compression_dim, n_rows, n_cols, scale, op
+    )
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:120]: Should be on one line

File: src/Compressors/Gaussian.jl [hunk 44]
```diff
@@ -0,0 +1,154 @@
...
+    randn!(S.op)
+    lmul!(S.scale, S.op)
+    return nothing
+end
+
+# Implement the matrix-Matrix Multiplication operators
+# Begin with the left version
+function mul!(
+    C::AbstractArray, S::GaussianRecipe, A::AbstractArray, alpha::Number, beta::Number
+)
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:139]: Arguments should be on their own lines

File: src/Compressors/Gaussian.jl [hunk 45]
```diff
@@ -0,0 +1,154 @@
...
+    left_mul_dimcheck(C, S, A)
+    # Built-in multiplication
+    mul!(C, S.op, A, alpha, beta)
+    return nothing
+end
+
+# Now implement the right versions
+function mul!(
+    C::AbstractArray, A::AbstractArray, S::GaussianRecipe, alpha::Number, beta::Number
+)
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:149]: Arguments should be on their own lines

File: src/Compressors/Gaussian.jl
Comment (vp314) [inline src/Compressors/Gaussian.jl:1]: Maybe this file name shouldb elower case to keep with the style of `sparse_sign.jl`

File: test/Compressors/Gaussian.jl
Comment (vp314) [inline test/Compressors/Gaussian.jl:1]: Should the file name be lower case?

File: test/Compressors/Gaussian.jl [hunk 46]
```diff
@@ -0,0 +1,287 @@
+module Gaussian_compressor
+using Test, RLinearAlgebra
+import Base.*
+import LinearAlgebra: mul!, lmul!
+import Random: randn!, seed!, rand
+using ..FieldTest
+using ..ApproxTol
+
+seed!(21321)
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:9]: Do we want to set the seed?

File: src/Compressors/Gaussian.jl [hunk 47]
```diff
@@ -0,0 +1,154 @@
...
+end
+
+function complete_compressor(ingredients::Gaussian, A::AbstractMatrix)
+    return GaussianRecipe(ingredients.cardinality, ingredients.compression_dim, ingredients.type, A)
+end
+
+# Allocations in this function are entirely due to bitrand call
+function update_compressor!(S::GaussianRecipe)
+    # Inplace update of sketch_matrix
+    randn!(S.op)
```
Comment (vp314) [inline src/Compressors/Gaussian.jl:130]: Do you need to pass the type here? Maybe not, just curious

File: test/Compressors/Gaussian.jl [hunk 48]
```diff
@@ -0,0 +1,287 @@
...
+            )
+        end
+
+        # Verify external constructor and type 
+        for Card in [Left, Right]
+            compressor = Gaussian(; cardinality=Card())
+            typeof(compressor.cardinality) == Card
+        end
+
+        for type in [Bool, Int16, Int32, Int64, Float16, Float32, Float64]
```
Comment (vp314) [inline test/Compressors/Gaussian.jl:41]: We probably do not want to test `Bool` or Integers, since this should not be allowed since we cannot generate a Gaussian random variable of this type.

**Review by vp314 — CHANGES_REQUESTED** [review]
Minor changes. Otherwise, the PR looks good. Some of the comments are about file naming conventions that should be discussed with Nathaniel.

File: src/Compressors/Gaussian.jl (hunk 47 above)
Comment (TongtongJin) [reply src/Compressors/Gaussian.jl:130]: i think randn!() should be able to in-place update the S.op as its own type.

File: test/Compressors/Gaussian.jl (hunk 48 above)
Comment (TongtongJin) [reply test/Compressors/Gaussian.jl:41]: Deleted Bool and integers.

File: test/Compressors/Gaussian.jl
Comment (TongtongJin) [reply test/Compressors/Gaussian.jl:1]: Changed into lower case

**Review by vp314 — APPROVED** [review]
